File: test/features/payments/history_screen_data_test.dart
import 'package:billpayments/core/mock/mock_network.dart';
import 'package:billpayments/features/auth/data/auth_provider.dart';
import 'package:billpayments/features/payments/data/payment.dart';
import 'package:billpayments/features/payments/data/payments_provider.dart';
import 'package:billpayments/features/payments/ui/history_screen_data.dart';
import 'package:billpayments/features/payments/ui/history_screen_data_provider.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('history is empty before any payment and newest-first after', () async {
    final container = ProviderContainer(overrides: [
      mockNetworkProvider.overrideWithValue(
          MockNetwork(minDelayMs: 0, maxDelayMs: 1, failEvery: 2)),
    ]);
    addTearDown(container.dispose);
    final auth = container.read(authProvider.notifier);
    await auth.sendOtp('9876543210');
    await auth.verifyOtp('123456');

    expect(container.read(historyScreenDataProvider),
        const HistoryScreenData.empty());

    final payments = container.read(paymentsProvider.notifier);
    await payments.pay(
        billerId: 'electricity-metro',
        billerName: 'Metro Electricity',
        categoryId: 'electricity',
        account: 'K1',
        amountPaise: 10000);
    await payments.pay(
        billerId: 'water-city',
        billerName: 'City Water',
        categoryId: 'water',
        account: 'W1',
        amountPaise: 20000);

    final loaded = container.read(historyScreenDataProvider) as HistoryLoaded;
    expect(loaded.items.length, 2);
    expect(loaded.items.first.billerName, 'City Water');
    expect(loaded.items.first.status, PaymentStatus.failed);
    expect(loaded.items.last.status, PaymentStatus.success);
    expect(loaded.items.first.amountPaise, 20000);
  });
}

File: __tests__/features/payments/historyScreenData.test.ts
import { projectHistory } from '@/features/payments/ui/historyScreenData';
import { emptyPayments, adding, type Payment } from '@/features/payments/data/payment';

const mk = (id: string, name: string, amount: number, status: Payment['status']): Payment => ({
  id, billerId: 'b', billerName: name, categoryId: 'c', account: 'a',
  amountPaise: amount, paidAtUtc: new Date(0), status,
});

describe('projectHistory', () => {
  it('empty before any payment', () => {
    expect(projectHistory(emptyPayments())).toEqual({ kind: 'empty' });
  });
  it('newest-first', () => {
    let p = adding(emptyPayments(), mk('pay-1', 'Metro Electricity', 10000, 'success'));
    p = adding(p, mk('pay-2', 'City Water', 20000, 'failed'));
    const r = projectHistory(p);
    expect(r.kind === 'loaded' && r.items[0]!.billerName).toBe('City Water');
    expect(r.kind === 'loaded' && r.items[0]!.status).toBe('failed');
    expect(r.kind === 'loaded' && r.items[1]!.status).toBe('success');
  });
});
